src/main/java/fi/aalto/cs/apluscourses/ui/APlusAuthenticationView.java

```diff
@@ -65,11 +63,30 @@ protected JComponent createCenterPanel() {
   protected ValidationInfo doValidate() {
     if (inputField.getPassword().length == 0) {
       return new ValidationInfo(getText("ui.authenticationView.noEmptyToken"),
-          inputField);
+          inputField).withOKEnabled();
+    }
+    try {
+      var input = inputField.getPassword();
+      authenticationViewModel.setToken(input);
+      Arrays.fill(input, '\0');
+      var authentication = authenticationViewModel.build();
+      authenticationViewModel.getUserName(authentication);
+    } catch (UnexpectedResponseException e) {
+      return new ValidationInfo(getText("ui.authenticationView.invalidToken"),
+          inputField).withOKEnabled();
+    } catch (IOException e) {
+      return new ValidationInfo(getText("ui.authenticationView.connectionError"),
+          inputField).withOKEnabled();
     }
+    super.doOKAction();
     return null;
   }
```

Comment:
Did you take a look at `OurDialogWrapper` and considered moving this logic to the view model? You can add a method call to something like `preValidation` (add an empty protected method to the class) in the beginning of `doValidate` of that class. Then you can override `preValidation` in this class and put lines 69-71 there. What do you think?

Reply:
Doing it this way seemed simpler to me, but I can change it if you want

Reply:
It is "simpler" this way, but the thing is unit-testability. We can't unit-test in view classes (or even if we could, we don't want to -- there are too many things we cannot control!). So pulling the validation logic to the view model layer is the solution. But there's no need to do that in this PR.